File: Python/aula12.py
try:
    resultado = 10/0
except ZeroDivisionError:
    print("Erro: Divisão por zero não é permitida")

lista = [1, 2, 3]

try:
    elemento = lista[5]
except IndexError:
    print("Erro: índice fora dos limites da lista")

dicionario = {"a":1, "b":2,"c":3, "d":4}

try:
    elemento = dicionario["e"]
except KeyError:
    print("Erro: essa chave não existe no dicionário")

def main():
    try:
        num1 = float(input("Digite o primeiro número: "))
        num2 = float(input("Digite o segundo número: "))

        resultado = num1/num2

    except ZeroDivisionError as e:
        print("Erro: divisão por zero não permitida")
    except ValueError as e:
        print("Vai dividir uma letra mano???")
    except Exception as e:
        print(f"Ocorreu um erro inesperado: {e}")
    else:
        print(f"O resultado da divisão é {resultado}")
    finally:
        print("programa finalizado!")

main()

string = "1"
try:
    numero = int(string)
except ValueError:
    print("Erro: conversão de string para inteiro falhou")
else:
    print("Conversão feita com sucesso")

dicionario = {"a":[1,2,3], "b":[4,5,6]}
try:
    elemento = dicionario["c"][0]
except KeyError:
    print("Erro: Chave inexistente no dicionario")
except IndexError:
    print("Erro: Índice fora dos limites da lista")

File: R/aula12.R
tryCatch({
    resultado <- 10/0
}, error = function(e){
    message("Erro: Divisão por zero não é permitida")
})

lista <- c(1,2,3)
tryCatch({
    elemento <- lista[5]
},error = function(e){
    message("Erro: índice fora dos limites da lista")
})

dicionario <- list(a=1, b=2, c=3, d=4)
tryCatch({
    elemento <- dicionario[["e"]]
},error = function(e){
    message("Erro: essa chave não existe no dicionário")
})


log_calculator <- function(x){
    tryCatch(
        expr = {
            message(log(x))
            message("Successfully executed the log(x) call.")
        },
        error = function(e){
            message('Caught an error!')
            print(e)
        },
        warning = function(w){
            message('Caught an warning!')
            print(w)
        },
        finally = {
            message('All done, quitting.')
        }
    )    
}

log_calculator(10)

string <- "abc"

tryCatch(
    expr ={
        numero <- as.integer(string)
}, warning = function(w){
    print("Erro: conversão de string para inteiro falhou")
})

dicionario <- list(a = c(1,2,3), b = c(4,5,6))
tryCatch({
    elemento <- dicionario[["c"]][1]
}, error = function(e){
    if(inherits(e, "simpleError")){
        print("Error: chave inexistente no dicionario ou indice fora dos limites da lista")
    }
})